src/thunderbots/software/util/visualizer_messenger/visualizer_messenger.cpp

```diff
@@ -0,0 +1,171 @@
+#include "visualizer_messenger.h"
+
+#include "util/constants.h"
+#include "util/logger/init.h"
+
+namespace Util
+{
+    std::shared_ptr<VisualizerMessenger> VisualizerMessenger::getInstance()
+    {
+        static std::shared_ptr<VisualizerMessenger> vm(new VisualizerMessenger);
+        return vm;
+    }
+
+    void VisualizerMessenger::initializePublisher(ros::NodeHandle node_handle)
+    {
+        if (m_publisher)
+        {
+            LOG(WARNING) << "Re-initializing visualizer messenger singleton publisher";
+        }
+
+        m_publisher = node_handle.advertise<LayerMsg>(
+            Util::Constants::VISUALIZER_DRAW_LAYER_TOPIC, 8);
+    }
+
+    const LayerMsgMap& VisualizerMessenger::getLayerMap() const
+    {
+        return m_layers_name_to_msg_map;
+    }
+
+    void VisualizerMessenger::update()
+    {
+        // Check if publisher is initialized before publishing messages
+        if (!m_publisher)
+        {
+            LOG(WARNING)
+                << "Update requested when visualizer messenger publisher is not initialized";
+            clearLayers();
+            return;
+        }
+
+        // Send layer messages
+        for (const std::pair<std::string, LayerMsg>& layer_msg_pair : getLayerMap())
+        {
+            if (!layer_msg_pair.second.shapes.empty())
+            {
+                m_publisher.publish(layer_msg_pair.second);
+            }
+        }
```

Comment:
Do we not clear the layers if the publisher was initialized? I would imagine it should be the other way around (ie. we clear the layers if the publisher **has been** initialized, and do not clear them if it has not been)?

Reply:
Oops, that's a bug. Should be fixed in a new push